sound/soc/sof/pcm.c

```diff
@@ -258,6 +260,8 @@ static int sof_pcm_hw_free(struct snd_pcm_substream *substream)
 	if (ret < 0)
 		dev_err(sdev->dev, "error: platform hw free failed\n");
 
+	spcm->prepared[substream->stream] = false;
+
```

Comment:
@kv2019i What happens in the "return ret" case at line 257 above? We abort freeing and leave the stream in a prepared state? But after `snd_pcm_lib_free_pages()` the stream probably won't work anymore either?

Reply:
@lyakh That's really outside this patches scope, I'm not touching that logic. To me it looks that if (!spcm) check fails on line 257, the function has been called with invalid arguments and we cannot proceed (the substream is not correctly set up).

Reply:
(!spcm) cannot fail since we dereference spcm above in cancel_work_sync(). Sorry, I meant a different line - I meant "return ret" at the *right* 257 / left 255, i.e. when IPC fails. I understand and agree, that it's out of scope of this patch set, but maybe it should be double-checked, that in that case we shouldn't call snd_sof_pcm_platform_hw_free(). As you know cleanup paths often ignore errors and actually cleanup functions often return void, so, they often just do as much cleanup as they can... Returning early from a cleanup isn't very common, unless some init steps haven't been completed earlier.